src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp

```diff
@@ -1569,6 +1569,355 @@ void C2_MacroAssembler::vinsert(BasicType typ, XMMRegister dst, XMMRegister src,
   }
 }
 
+void C2_MacroAssembler::vpackI2X(BasicType elem_bt, XMMRegister dst,
+                                 XMMRegister ones, XMMRegister xtmp,
+                                 int vlen_enc) {
+  assert(VM_Version::supports_avx512vl(), "");
+  if (elem_bt == T_SHORT) {
+    evpmovdw(dst, dst, vlen_enc);
+  } else {
+    assert(elem_bt == T_BYTE, "");
+    evpmovdb(dst, dst, vlen_enc);
+  }
+}
+
+void C2_MacroAssembler::vgather_subword_avx3(BasicType elem_bt, XMMRegister dst, Register base, XMMRegister offset,
+                                             Register idx_base, int idx_off, XMMRegister idx_vec, XMMRegister ones,
+                                             XMMRegister xtmp, KRegister gmask, int vlen_enc) {
+  int mask_shift = -1;
+  int bitlevel_offset_shift = -1;
+  int nomlarized_index_shift = -1;
+  Address::ScaleFactor scale = Address::no_scale;
+
+  if (elem_bt == T_SHORT) {
+    mask_shift = 31;
+    scale = Address::times_2;
+    bitlevel_offset_shift = 4;
+    nomlarized_index_shift = 1;
+  } else {
+    assert(elem_bt == T_BYTE, "");
+    mask_shift = 30;
+    scale = Address::times_1;
+    bitlevel_offset_shift = 3;
+    nomlarized_index_shift = 2;
+  }
+
+  evmovdquq(idx_vec, Address(idx_base, idx_off, Address::times_4), vlen_enc);
+  if (offset != xnoreg) {
+    vpaddd(idx_vec, idx_vec, offset, vlen_enc);
+  }
+  // Normalize the indices to multiple of nomlarized_index_shift.
+  vpslld(xtmp, ones, nomlarized_index_shift, vlen_enc);
+  vpand(xtmp, idx_vec, xtmp, vlen_enc);
+  // Load double words from normalized indices.
+  evpgatherdd(dst, gmask, Address(base, xtmp, scale), vlen_enc);
```

Comment:
Could we not do here directly:
evpgatherdd(dst, gmask, Address(base, idx_vec, scale), vlen_enc);
Then we dont need lines 1609-1611 and also 1616-1621 as well.

Reply:
Gathering is happening at doubleword granularity, where as masks are applier over sub-word lanes.